### DISCUSSION TITLE
Feature Request - Guest Authentication

### DISCUSSION THREAD

**Discussion by manishonc** [pocketbase/pocketbase#3035] [Q&A] [answered]
Please add a Guest Auth or Anonymous Auth.

In My use case the user doesn't need to register until he wants 
But . I want to give users some personalized experience within the app .

But with User Id I am not able to do that.
If there are any alternative solutions please advise me.

Thanks @ganigeorgiev

**Comment by ganigeorgiev** [comment, marked as answer]
This was discussed several times previously and there no plans to add dedicated "guest authentication" because this can be done already with the existing APIs, although it may not be as straightforward as in Firebase for example.

You can create the user entry behind the scene with some random generated password (_with maybe some special flag in the `users` collection_) and store the created user and password client-side.

When the user decide to finally "create/register" their account, you can perform an `update()` for the client-side stored user with the new password that the user explicitly set.

**Reply by manishonc** [reply]
Thanks

**Reply by aonghas** [reply]
Just checking... are there performance implications for creating many users simultaneously (different to regular base collection)? I'm thinking large scale events where hundreds of people scan a QR code to join an event and a "guest" user is created by every device at the same time upon page load...

**Reply by lukepighetti** [reply]
@ganigeorgiev is there any way to convert a username/password created user account to use oauth (like Sign in with Apple)?

**Reply by ganigeorgiev** [reply]
@lukepighetti The linking should be automatically done if the OAuth2 auth request is made by an authenticated user. For example:

```js
await pb.collection("users").authWithPassword("example", "1234567890")

// if authenticating for the first time with Apple,
// PocketBase will try to link the Apple account to the already authenticated "example" user
// (if not already linked to another user)
await pb.collection("users").authWithOAuth2({ provider: "apple" })
```

**Reply by ganigeorgiev** [reply]
_@lukepighetti Side-note: keep in mind that the Apple OAuth2 implementation is currently limited and may not work if you need to fetch the user email (see https://github.com/pocketbase/pocketbase/discussions/3814; I don't have Apple developers account to verify it and I'm not planning investing time on this but as mentioned in the discussion, I'll accept a PR if the required changes are not so drastic)._

**Reply by lukepighetti** [reply]
@ganigeorgiev what's the best way to determine if a user record is "anonymous" or "upgraded" to OAuth?

**Reply by ganigeorgiev** [reply]
@lukepighetti Hm, I'm not sure. It depends on your desired app flow. The simplest way to distinguish them is to check the `verified` state (_for username+pass only users it will be always `false` and the `verified` state should automatically change to `true` after OAuth2 auth_). Another option could be to add your own `bool` type field in the `users` collection but it will require also extra care to prevent users from manually editing it (eg. with API rules or hooks).

**Reply by ganigeorgiev** [reply]
Another option to determine whether the user record is "anonymous" or "upgraded" is to check client-side whether you have information about their password. If Yes, then the user could be considered "guest", otherwise regular user (eg. if the locally generated password is cleared after OAuth2 auth).

**Reply by laurentlahmy** [reply]
@ganigeorgiev, thank you so much for PocketBase.

It seems to me that verified being set to true after OAuth2 auth is not correct (anymore?).

> @lukepighetti Hm, I'm not sure. It depends on your desired app flow. The simplest way to distinguish them is to check the `verified` state (_for username+pass only users it will be always `false` and the `verified` state should automatically change to `true` after OAuth2 auth_). Another option could be to add your own `bool` type field in the `users` collection but it will require also extra care to prevent users from manually editing it (eg. with API rules or hooks).

my flow / steps to reproduce:

```tsx

const id = nanoid();
const create_guest_user = await pb.collection("users").create({
  email: `guest@${id}.com`,
  password: id,
  passwordConfirm: id,
});

const authenticate_guest_user = await pb
  .collection("users")
  .authWithPassword(`guest@${id}.com`, id);
  
const auth_with_google = await pb
  .collection("users")
  .authWithOAuth2({ provider: "google" });

```

 the authentication seems to work, it returns: 
 
```json

{
    "meta": {
        "accessToken": "ya29.a0AS3H6Nx[REDACTED]",
        "avatarURL": "https://lh3.googleusercontent.com/a/[REDACTED]",
        "avatarUrl": "https://lh3.googleusercontent.com/a/[REDACTED]",
        "email": "[REDACTED]@gmail.com",
        "expiry": "2025-07-10 17:37:30.640Z",
        "id": "[REDACTED]",
        "isNew": false,
        "name": "[REDACTED]",
        "rawUser": {
            "email": "[REDACTED]@gmail.com",
            "email_verified": true,
            "family_name": "[REDACTED]",
            "given_name": "[REDACTED]",
            "name": "[REDACTED]",
            "picture": "https://lh3.googleusercontent.com/a/[REDACTED]",
            "sub": "[REDACTED]"
        },
        "refreshToken": "",
        "username": ""
    },
    "record": {
        "avatar": "",
        "collectionId": "_pb_users_auth_",
        "collectionName": "users",
        "created": "2025-07-10 16:22:09.817Z",
        "email": "guest@[REDACTED].com",
        "emailVisibility": false,
        "guestId": "[REDACTED]",
        "id": "[REDACTED]",
        "meta": {
            "accessToken": "ya29.a0AS3H6Nw[REDACTED]",
            "avatarURL": "https://lh3.googleusercontent.com/a/[REDACTED]",
            "avatarUrl": "https://lh3.googleusercontent.com/a/[REDACTED]",
            "email": "[REDACTED]@gmail.com",
            "expiry": "2025-07-10 17:23:16.455Z",
            "id": "[REDACTED]",
            "isNew": false,
            "name": "[REDACTED]",
            "rawUser": {
                "email": "[REDACTED]@gmail.com",
                "email_verified": true,
                "family_name": "[REDACTED]",
                "given_name": "[REDACTED]",
                "name": "[REDACTED]",
                "picture": "https://lh3.googleusercontent.com/a/[REDACTED]",
                "sub": "[REDACTED]"
            },
            "refreshToken": "",
            "username": ""
        },
        "name": "",
        "updated": "2025-07-10 16:37:32.837Z",
        "verified": false
    },
    "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.[REDACTED]"
}

```

and the user is "merged", no new user is created, but no fields are updated.

User created as guest, authenticated with email & password after OAuth2 with provider Google (you can ignore `guestId` and the `meta` fields that I update manually): 
<img width="1453" height="91" alt="image" src="https://github.com/user-attachments/assets/e1e7b228-cc16-4a6c-bd5d-52f73701784a" />

[PocketBase v0.28.2]

**Reply by ganigeorgiev** [reply]
@laurentlahmy If the OAuth2 email doesn't match with the custom user submitted one, then No - it will not mark the account as verified and this is working as intended.

As a side-note and in relation to the original discussion, note that email doesn't need to be required or even submitted when creating a guest user.

**Side-note2: Creating random emails with domain that it is not controlled by you or IANA should be discouraged because it will be a vulnerability if an attacker register the domain.**

**Reply by laurentlahmy** [reply]
thank you.

I got it to work by following these steps: 

in the `users collection`
- remove nonEmpty on the email field
- make a guestId field
- make the guestId unique
- go to `Edit Collection` (settings icon) > `Options` > `Auth methods` > `Identity/Password` > set `Unique identity fields` to `guestId`
 
 ```jsx
const create_guest_user = await pb.collection("users").create({
  guestId: id,
  password: id,
  passwordConfirm: id,
});

const authenticate_guest_user = await pb
  .collection("users")
  .authWithPassword(id, id);
  
 const auth_with_google = await pb
  .collection("users")
  .authWithOAuth2({ provider: "google" });
```

**Reply by moritz4** [reply]
Has the option to 'remove nonEmpty on the email field'  been removed? Setting the min length to blank doesn't seem to work either.

<img width="659" height="290" alt="image" src="https://github.com/user-attachments/assets/828a3275-5b2a-48ff-a173-31570ebeb916" />

**Reply by ganigeorgiev** [reply]
@moritz4 No, it is still there. The above screenshot shows the `password`, not the `email` field.
